### PR TITLE
[Catalog] Add GithubApiReaderProcessor #1525

### PR DESCRIPTION
**Pull request by dandehoon**
Add Processor that uses Github API v3 to support importing private Github repositories.

Reference: 
[https://docs.github.com/en/rest/reference/repos#contents](https://docs.github.com/en/rest/reference/repos#contents)

#### :heavy_check_mark: Checklist

- [x] All tests are passing `yarn test`
- [ ] Screenshots attached (for UI changes)
- [ ] Relevant documentation updated
- [x] Prettier run on changed files
- [x] Tests added for new functionality
- [ ] Regression tests added for bug fixes

### CODE REVIEW

**Comment by andrewthauer** [comment]
🎉 This should address #1525

File: plugins/catalog-backend/src/ingestion/processors/GithubReaderProcessor.ts [hunk 1]
```diff
@@ -15,11 +15,29 @@
  */
 
 import { LocationSpec } from '@backstage/catalog-model';
-import fetch from 'node-fetch';
+import fetch, { RequestInit, HeadersInit } from 'node-fetch';
 import * as result from './results';
 import { LocationProcessor, LocationProcessorEmit } from './types';
 
 export class GithubReaderProcessor implements LocationProcessor {
+  private privateToken: string = process.env.GITHUB_PRIVATE_TOKEN || '';
```
Comment (Rugvip) [inline plugins/catalog-backend/src/ingestion/processors/GithubReaderProcessor.ts:23]: Managing this token is really tricky, and I think the only "real" way that GitHub provides to do this atm is via GitHub App installations. @benjdlambert is looking into how we could support that in the best way, but in the meantime I think it makes sense to go forward with this.

**Review by Rugvip — COMMENTED** [review]
Thanks! 😁 

I'm assuming that this adds a requirement to set up a GitHub token to be able to add any GitHub locations to the catalog now right? I think the best would be if we add this as a separate `GithubApiReaderProcessor`, so that we can support both raw and api reading. The default one would be the raw one, but it would make it simpler to set up a custom deployment with one that uses the API.

**Comment by danztran** [comment]
> Thanks! 😁
> 
> I'm assuming that this adds a requirement to set up a GitHub token to be able to add any GitHub locations to the catalog now right? I think the best would be if we add this as a separate `GithubApiReaderProcessor`, so that we can support both raw and api reading. The default one would be the raw one, but it would make it simpler to set up a custom deployment with one that uses the API.

I think it's better with only API Reading. Because Raw Reading fetches from `raw.githubusercontent.com` in the second call will hit a cached version of the file, but API Reading won't. And in case the token is not provided, API Reading still can fetch from Public Repositories too.

**Comment by lndbrg** [comment]
I agree with @danztran, prefer always going the api route and it's quite easy to translate raw to api requests. Back in the days the service catalog reader at spotify took the html uri and replaced it with the raw uri on the fly. As long as you know the base uri for the api endpoint (should probably be configurable) the rewrite should be fairly easy to do. But probably nudge the user in the right direction.

File: plugins/catalog-backend/src/ingestion/processors/GithubReaderProcessor.ts (hunk 1 above)
Comment (benjdlambert) [reply plugins/catalog-backend/src/ingestion/processors/GithubReaderProcessor.ts:23]: There's a shiny ✨  RFC here: https://github.com/spotify/backstage/issues/1565

**Comment by Rugvip** [comment]
Yeah I agree API is better and I know that GitHub themselves prefer that people use it over raw, but I'm not sure it's feasible in this case. Calling the API without a token isn't very useful, as you're limited to 60 req/h, which is consumed real quick with the 8 example entities that we have. I think the best would be to keep both processors around, and then look into if we can remove the raw one as we bring in support for Github Apps.

**Review by Rugvip — APPROVED** [review]
Sick, thanks! 😁

**Comment by benjdlambert** [comment]
@danztran would be good to get some documentation written for how to use this, but then realised we don't have any documentation about how to load fixtures with the other way, so I'll create another ticket to fix that too and we can fix that.

Obviously it might change pretty soon with the outcome of the RFC #1565. So will probably leave it for now.

Nice work 🍰 🏅

**Pull request merged**